File: binarysearch.cpp
#include <bits/stdc++.h>
using namespace std;

class IceCream {
    
    public: 
        int flavor; 
        int index;

        IceCream(int flavor, int index) {
            
            this->flavor=flavor;
            this->index=index;
            
       }
};
    
int binarySearch(int first, int last, vector<IceCream> arr, int search) {
    
    while(first<last){
        int middle = (first+last/2)-1;
        
        if(arr[middle].flavor==search){return arr[middle].index;}
        
        else if(arr[middle].flavor<search){first=middle+1;}
        
        else if(arr[middle].flavor>search){last=middle-1;}
        
    }
    
    return -1;
   
}

bool compare(IceCream a, IceCream b){
    
    return a.flavor < b.flavor;
}

int main() {
    int t;
    int n, m;
    cin >> t;
    for(int test = 0; test < t; test++) {       
        cin >> m >> n;
        vector<IceCream> arr;
        arr.reserve(n); 

        for (int i = 0; i < n; i++) {
            int cost;
            cin >> cost;
            arr.push_back(IceCream(cost, i + 1));
        }

        sort(arr.begin(), arr.end(), compare);
        
        int firstIndex = 100000, secondIndex = 100000;
        for(int i = 0; i < n - 1 ; i++) {
            //cout << arr[i].flavor;
            int search = m - arr[i].flavor;
            if(search >= arr[i].flavor) {
                int index = binarySearch( i + 1, n - 1, arr, search);
                if( index != -1 ) {
                    cout << min(arr[i].index, index) << " " << max(arr[i].index, index) << endl;
                    break;

                }
            }
        }

    }

}

File: binary-search.py
import numpy as np


n=100
random_list = np.random.rand(10)*n

random_int = random_list[7]

random_list.sort()
print(random_list)

def binary_check(num,random_list):

    while (len(random_list) > 0):
        x = len(random_list)//2
        if(random_list[x] == num):
            #print(random_list[x])
            return 1
        elif(random_list[x] > num):
            random_list = random_list[0:x]
        elif(random_list[x] < num):
            random_list = random_list[x:]

    return 0


def binary_search(num,random_list,n):
    first = 1
    last = len(random_list)

    while(first<=last):
        middle = (first+last//2)

        if(random_list[middle] == num):
            return middle
        elif(random_list[middle] > num):
            last = middle-1
        elif(random_list[middle] < num):
            first = middle+1

print(random_int)
#print(binary_check(random_int,random_list))
print(binary_search(random_int,random_list,n))

'''
a = 0
b = n-1
x = a+b/2
'''
